libs/geo/src/test/java/org/elasticsearch/geo/geometry/BaseGeometryTestCase.java

```diff
@@ -0,0 +1,203 @@
+        Object result = geom.visit(new GeometryVisitor<Object>() {
+            private Object verify(Geometry geometry, String expectedClass) {
+                assertFalse("Visitor should be called only once", called.getAndSet(true));
+                assertSame(geom, geometry);
+                assertEquals(geometry.getClass().getName(), "org.elasticsearch.geo.geometry." + expectedClass);
+                return "result";
+            }
+
+            @Override
+            public Object visit(Circle circle) {
+                return verify(circle, "Circle");
+            }
+
+            @Override
+            public Object visit(GeometryCollection<?> collection) {
+                return verify(collection, "GeometryCollection");            }
+
+            @Override
+            public Object visit(Line line) {
+                return verify(line, "Line");
+            }
+
+            @Override
+            public Object visit(LinearRing ring) {
+                return verify(ring, "LinearRing");
+            }
+
+            @Override
+            public Object visit(MultiLine multiLine) {
+                return verify(multiLine, "MultiLine");
+            }
+
+            @Override
+            public Object visit(MultiPoint multiPoint) {
+                return verify(multiPoint, "MultiPoint");
+            }
+
+            @Override
+            public Object visit(MultiPolygon multiPolygon) {
+                return verify(multiPolygon, "MultiPolygon");
+            }
+
+            @Override
+            public Object visit(Point point) {
+                return verify(point, "Point");
+            }
+
+            @Override
+            public Object visit(Polygon polygon) {
+                return verify(polygon, "Polygon");
+            }
+
+            @Override
+            public Object visit(Rectangle rectangle) {
+                return verify(rectangle, "Rectangle");
+            }
+        });
+
+        assertTrue("visitor wasn't called", called.get());
+        assertEquals("result", result);
+    }
+
+    public static double randomLat() {
+        return randomDoubleBetween(-90, 90, true);
+    }
+
+    public static double randomLon() {
+        return randomDoubleBetween(-180, 180, true);
+    }
+
+    public static Circle randomCircle() {
+        return new Circle(randomDoubleBetween(-90, 90, true), randomDoubleBetween(-180, 180, true), randomDoubleBetween(0, 100, false));
+    }
+
+    public static Line randomLine() {
+        int size = randomIntBetween(2, 10);
+        double[] lats = new double[size];
+        double[] lons = new double[size];
+        for (int i = 0; i < size; i++) {
+            lats[i] = randomLat();
+            lons[i] = randomLon();
+        }
+        return new Line(lats, lons);
+    }
+
+    public static Point randomPoint() {
+        return new Point(randomLat(), randomLon());
+    }
+
+    public static LinearRing randomLinearRing() {
```

Comment:
I think we should reuse lucene's `GeoTestUtil` for creating these random geometries. There was a lot of work put in to making sure the test suite creates adversarial geometries that have a tendency to expose weaknesses in the API. I don't think that should hold up the PR, though. So lets create a separate issue to migrate some of the random geometry generators over to lucene's geo test suite.

Reply:
I opened #37278 as a follow-up.